File: BeeCrowd/3247.cpp
#include <iostream>
#include <queue>
#include <cmath>
#include<map>
using namespace std;

#define debug(x) cout<<#x<<"= "<<x<<endl
#define mod(x) ((x)>0)?(x):(-(x))
#define MAX(a,b) ((a)>(b))? (a):(b)
#define MIN(a,b) ((a)<(b))? (a):(b)

class node{
public:
	int x;
	int y;

	node(int x, int y)
	{
		this->x = x;
		this->y = y;
	}
};

bool insideGrid (int rows, int cols, int x, int y){
	if (x>=0 && x<rows && y>=0 && y<cols) return true;
	else return false;
}

void push_node (int x, int y, const vector<vector<long long> > &grid,vector<vector<bool> > &vis, queue<node> &q){
	if (insideGrid(grid.size(), grid.size(), x, y) && grid[x][y]>=0 && !vis[x][y]){
		vis[x][y] = true;
		q.push(node(x, y));
	}
}

int bfs(vector<vector<long long> > &grid){
	int n = grid.size();
	queue<node> q;
	vector<vector<bool> > vis;

	for (int i=0; i<n; i++){
		vis.push_back(vector<bool>());
		for (int j=0; j<n; j++) vis[i].push_back(false);
	}

	q.push (node(0, 0));
	vis[0][0] = true;
	while (!q.empty()){
    int x, y;
		node popNode = q.front();
		q.pop();

		if (popNode.x==n-1 && popNode.y==n-1) return true;

    //direita
		x = popNode.x;
		y = popNode.y+1;
		push_node(x, y, grid, vis, q);
    //esquerda
		x = popNode.x-1;
		y = popNode.y;
		push_node(x, y, grid, vis, q);
    //cima
		x = popNode.x+1;
		y = popNode.y;
		push_node(x, y, grid, vis, q);
    //baixo
		x = popNode.x;
		y = popNode.y-1;
		push_node(x, y, grid, vis, q);
	}
	return false;
}

void dynamicMatrix(vector<vector<long long> > &grid){
	int size = grid.size();
	long long modulo = (1LL<<31) -1;
  bool flag = false;

  // coluna 0
	for (int i=0; i<size; i++){
    if (grid[i][0] == 0) grid[i][0] = 1;
		else break;
  }
  // linha 0
	for (int i=1; i<size; i++){
    if (grid[0][i] == 0)grid[0][i]=1;
		else break;
  }
		
	for (int i=1; i<size; i++){
    for (int j=1; j<size; j++){
			if (grid[i][j]==-1)continue;
			if (grid[i-1][j] > 0) grid[i][j] = (grid[i][j] + grid[i-1][j]) % modulo;
			if (grid[i][j-1] > 0) grid[i][j] = (grid[i][j]+ grid[i][j-1]) % modulo;
		}
  }
}

int main (){
	int n;
	cin>>n;
	vector<vector<long long> > grid;
	for (int i=0; i<n; i++){
		grid.push_back(vector<long long>());
		for (int j=0; j<n; j++){
			char c;
			cin>>c;
			if (c == '#') grid[i].push_back(-1);
			else grid[i].push_back(0);
		}
	}

	dynamicMatrix(grid);
	if (grid[n-1][n-1] != 0){
    cout<<grid[n-1][n-1]<<endl;
  }else{
		if (!bfs(grid)) cout<<"INCONCEIVABLE"<<endl;
		else cout<<"THE GAME IS A LIE"<<endl;
	} 
}

File: BeeCrowd/3247.py
from collections import deque

def dpPath(obstacleGrid):
  rows = len(obstacleGrid)
  cols = len(obstacleGrid[0])
  dp = list(list(0 for i in range(cols)) for i in range (rows))
  flag = False
  
  # first row
  for i in range (cols):
    if flag or obstacleGrid[0][i] == 1:
      flag = True
      dp[0][i] = 0
    else:
      dp[0][i] = 1
      
  # first column
  flag = False
  for i in range (rows):
    if flag or obstacleGrid[i][0] == 1:
      flag = True
      dp[i][0] = 0
    else:
      dp[i][0] = 1
      
  for i in range(1,rows):
    for j in range(1,cols):
      if(obstacleGrid[i][j] == 1):
        dp[i][j] = 0
      else:
        dp[i][j] = (dp[i-1][j] + dp[i][j-1]) 
      
        
  return dp[rows-1][cols-1]

def bfs(grid):
  R, C = len(grid), len(grid[0])
  q = deque()
  vis = [[False]*R for i in range(C)]
  
  q.append((0,0))
  vis[0][0] = True
  while len(q):
    x, y = q.popleft()
  
    if(x,y) == (R-1,C-1): return True
    
    for dx, dy in [(-1, 0), (1, 0), (0, -1), (0, 1)]:
      new_x, new_y = x+dx, y+dy
      if 0 <= new_x < R and 0 <= new_y < C and not vis[new_x][new_y] and grid[new_x][new_y] == 0:
        vis[new_x][new_y] = True
        q.append((new_x, new_y))
        
  return False

# MAIN

grid = list()
n = int(input())
for i in range(n):
  s = input()
  s = s.replace(".", "0")
  s = s.replace("#", "1")
  line = list(int(char) for char in s)
  grid.append(line)
  
dpResult = dpPath(grid)  
  
if dpResult != 0:
  print(dpResult)
else:
  if bfs(grid): print("THE GAME IS A LIE")
  else: print("INCONCEIVABLE")
